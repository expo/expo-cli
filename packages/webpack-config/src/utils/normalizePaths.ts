/**
 * Converts absolute paths to relative paths for testing purposes.
 *
 * @param initial
 * @param transformString
 * @internal
 */
export default function normalizePaths(
  initial: any,
  transformString: (value: string) => string
): any {
  if (initial == null) {
    return initial;
  } else if (typeof initial === 'string') {
    return transformString(initial);
  } else if (Array.isArray(initial)) {
    return initial.map(value => normalizePaths(value, transformString));
  } else if (typeof initial === 'object') {
    let result: { [key: string]: any } = {};
    for (const prop of Object.keys(initial)) {
      result[prop] = normalizePaths(initial[prop], transformString);
    }
    return result;
  } else {
    return initial;
  }
}
