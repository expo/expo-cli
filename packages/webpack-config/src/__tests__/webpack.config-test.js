import { dirname } from 'path';

import createConfig from '..';
import normalizePaths from '../utils/normalizePaths';

jest.unmock('resolve-from');

const projectRoot = dirname(require.resolve('@expo/webpack-config/e2e/basic'));

function normalizeConfig(config) {
  // Replace plugins with their names
  // in the future we may want to selectively test certain plugin shapes
  const plugins = config.plugins;
  config.plugins = [];
  // Just use plugin names in order
  for (const plugin of plugins) {
    config.plugins.push(plugin.constructor.name);
  }

  // Make the paths be relative to the project
  const normalized = normalizePaths(config, value => value.split('cli/').pop());

  // Strip out any path issues that come from testing in a monorepo.
  function normalizeLoaders(loaders) {
    return loaders.map(rule => {
      if (rule.use) {
        rule.use = normalizePaths(rule.use, value => {
          if (value.includes('node_modules')) {
            return `node_modules` + value.split('node_modules').pop();
          }
          return value;
        });
      } else if (rule.oneOf) {
        rule.oneOf = normalizeLoaders(rule.oneOf);
      }
      return rule;
    });
  }

  normalized.module.rules = normalizeLoaders(normalized.module.rules);

  delete normalized.devServer?.watchOptions;

  // performance is disabled in CI
  delete normalized.performance;
  return normalized;
}

let originalCiValue;

beforeAll(() => {
  originalCiValue = process.env.CI;
  process.env.CI = 'true';
});

afterAll(() => {
  process.env.CI = originalCiValue;
});

describe(`ios`, () => {
  it('ios development', async () => {
    const config = await createConfig({ mode: 'development', platform: 'ios', projectRoot });
    const normalized = normalizeConfig(config);
    expect(normalized).toMatchSnapshot();
  });
  it('ios production', async () => {
    const config = await createConfig({ mode: 'production', platform: 'ios', projectRoot });
    const normalized = normalizeConfig(config);

    expect(normalized).toMatchSnapshot();
  });
});
describe(`web`, () => {
  // {offline: true} should add the copy webpack plugin and change the entry to account for the register-service-worker file.
  it('web offline', async () => {
    const config = await createConfig({
      mode: 'development',
      offline: true,
      platform: 'web',
      projectRoot,
    });
    const normalized = normalizeConfig(config);
    expect(typeof normalized.entry).toBe('function');
    expect(normalized.plugins).toContain('CopyPlugin');
  });

  it('web development', async () => {
    const config = await createConfig({ mode: 'development', platform: 'web', projectRoot });
    const normalized = normalizeConfig(config);

    expect(normalized).toMatchSnapshot();
  });

  it('web production', async () => {
    const config = await createConfig({ mode: 'production', platform: 'web', projectRoot });

    const normalized = normalizeConfig(config);

    expect(normalized).toMatchSnapshot();
  });
});
