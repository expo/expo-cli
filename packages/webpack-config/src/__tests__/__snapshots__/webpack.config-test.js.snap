// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ios ios development 1`] = `
Object {
  "bail": false,
  "context": "packages/webpack-config/e2e/basic",
  "devServer": Object {
    "after": [Function],
    "before": [Function],
    "clientLogLevel": "none",
    "compress": true,
    "contentBase": "packages/webpack-config/web-default",
    "contentBasePublicPath": "/",
    "disableHostCheck": true,
    "historyApiFallback": Object {
      "disableDotRule": true,
      "index": "/",
    },
    "host": "0.0.0.0",
    "hot": true,
    "https": false,
    "injectClient": false,
    "inline": false,
    "mimeTypes": Object {
      "force": true,
      "typeMap": Object {
        "application/javascript": Array [
          "bundle",
        ],
      },
    },
    "overlay": false,
    "proxy": undefined,
    "public": undefined,
    "publicPath": "/",
    "quiet": true,
    "sockHost": undefined,
    "sockPath": undefined,
    "sockPort": undefined,
    "transportMode": "ws",
    "watchContentBase": true,
  },
  "devtool": "cheap-module-source-map",
  "entry": Object {
    "app": Array [
      "node_modules/react-native/Libraries/polyfills/console.js",
      "node_modules/react-native/Libraries/polyfills/error-guard.js",
      "node_modules/react-native/Libraries/polyfills/Object.es7.js",
      "node_modules/react-native/Libraries/Core/InitializeCore.js",
      "packages/webpack-config/e2e/basic/index.js",
    ],
  },
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "parser": Object {
          "requireEnsure": false,
        },
      },
      Object {
        "oneOf": Array [
          Object {
            "exclude": "packages/webpack-config/web-default",
            "test": /\\\\\\.html\\$/,
            "use": Array [
              "node_modules/html-loader/dist/cjs.js",
            ],
          },
          Object {
            "test": /\\\\\\.\\(gif\\|jpe\\?g\\|png\\|svg\\)\\$/,
            "use": Object {
              "loader": "node_modules/url-loader/dist/cjs.js",
              "options": Object {
                "esModule": false,
                "limit": 1000,
                "name": "static/media/[name].[hash:8].[ext]",
              },
            },
          },
          Object {
            "include": [Function],
            "test": /\\\\\\.\\(mjs\\|\\[jt\\]sx\\?\\)\\$/,
            "use": Object {
              "loader": "node_modules/babel-loader/lib/index.js",
              "options": Object {
                "babelrc": false,
                "cacheCompression": false,
                "cacheDirectory": "packages/webpack-config/e2e/basic/.expo/ios/cache/development/babel-loader",
                "cacheIdentifier": "custom-value-to-skip-babel-config-error",
                "caller": Object {
                  "__dangerous_rule_id": "expo-babel-loader",
                  "bundler": "webpack",
                  "mode": "development",
                  "platform": "ios",
                },
                "compact": false,
                "configFile": true,
                "inputSourceMap": true,
                "root": "packages/webpack-config/e2e/basic",
                "sourceMaps": true,
                "sourceType": "unambiguous",
              },
            },
          },
          Object {
            "test": /\\\\\\.worker\\\\\\.\\(js\\|mjs\\|ts\\)\\$/,
            "use": Object {
              "loader": "node_modules/worker-loader/dist/cjs.js",
            },
          },
          Object {
            "include": Array [
              "packages/webpack-config/e2e/basic",
              "packages/webpack-config/e2e/basic/node_modules/react-native-vector-icons",
              "packages/webpack-config/e2e/basic/node_modules/@expo/vector-icons",
            ],
            "test": /\\\\\\.\\(woff2\\?\\|eot\\|ttf\\|otf\\)\\$/,
            "use": Array [
              Object {
                "loader": "node_modules/url-loader/dist/cjs.js",
                "options": Object {
                  "esModule": false,
                  "limit": 50000,
                  "name": "./fonts/[name].[ext]",
                },
              },
            ],
          },
          Object {
            "test": /\\\\\\.\\(css\\)\\$/,
            "use": Array [
              "node_modules/style-loader/dist/cjs.js",
              "node_modules/css-loader/dist/cjs.js",
            ],
          },
          Object {
            "exclude": Array [
              /\\\\\\.\\(mjs\\|\\[jt\\]sx\\?\\)\\$/,
              /\\\\\\.html\\$/,
              /\\\\\\.json\\$/,
            ],
            "loader": "node_modules/file-loader/dist/cjs.js",
            "options": Object {
              "esModule": false,
              "name": "static/media/[name].[hash:8].[ext]",
            },
          },
        ],
      },
    ],
    "strictExportPresence": false,
  },
  "output": Object {
    "chunkFilename": "static/js/[name].chunk.js",
    "devtoolModuleFilenameTemplate": [Function],
    "filename": "static/js/bundle.js",
    "globalObject": "this",
    "path": "packages/webpack-config/e2e/basic/web-build",
    "pathinfo": true,
    "publicPath": "/",
  },
  "plugins": Array [
    "HtmlWebpackPlugin",
    "InterpolateHtmlPlugin",
    "ExpoAppManifestWebpackPlugin",
    "ModuleNotFoundPlugin",
    "DefinePlugin",
    "LimitChunkCountPlugin",
    "WatchMissingNodeModulesPlugin",
  ],
  "resolve": Object {
    "extensions": Array [
      ".ios.ts",
      ".ios.tsx",
      ".ios.js",
      ".ios.jsx",
      ".native.ts",
      ".native.tsx",
      ".native.js",
      ".native.jsx",
      ".ts",
      ".tsx",
      ".js",
      ".jsx",
      ".json",
    ],
    "plugins": Array [
      Object {
        "apply": [Function],
        "bind": [Function],
        "forkTsCheckerOptions": [Function],
        "makePlugin": [Function],
        "moduleLoader": [Function],
        "topLevelLoader": Object {
          "apply": [Function],
        },
        "tsLoaderOptions": [Function],
      },
    ],
    "symlinks": false,
  },
  "resolveLoader": Object {
    "plugins": Array [
      Object {
        "apply": [Function],
      },
    ],
  },
  "target": "webworker",
}
`;

exports[`ios ios production 1`] = `
Object {
  "bail": true,
  "context": "packages/webpack-config/e2e/basic",
  "devtool": "source-map",
  "entry": Object {
    "app": Array [
      "node_modules/react-native/Libraries/polyfills/console.js",
      "node_modules/react-native/Libraries/polyfills/error-guard.js",
      "node_modules/react-native/Libraries/polyfills/Object.es7.js",
      "node_modules/react-native/Libraries/Core/InitializeCore.js",
      "packages/webpack-config/e2e/basic/index.js",
    ],
  },
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "parser": Object {
          "requireEnsure": false,
        },
      },
      Object {
        "oneOf": Array [
          Object {
            "exclude": "packages/webpack-config/web-default",
            "test": /\\\\\\.html\\$/,
            "use": Array [
              "node_modules/html-loader/dist/cjs.js",
            ],
          },
          Object {
            "test": /\\\\\\.\\(gif\\|jpe\\?g\\|png\\|svg\\)\\$/,
            "use": Object {
              "loader": "node_modules/url-loader/dist/cjs.js",
              "options": Object {
                "esModule": false,
                "limit": 1000,
                "name": "static/media/[name].[hash:8].[ext]",
              },
            },
          },
          Object {
            "include": [Function],
            "test": /\\\\\\.\\(mjs\\|\\[jt\\]sx\\?\\)\\$/,
            "use": Object {
              "loader": "node_modules/babel-loader/lib/index.js",
              "options": Object {
                "babelrc": false,
                "cacheCompression": false,
                "cacheDirectory": "packages/webpack-config/e2e/basic/.expo/ios/cache/production/babel-loader",
                "cacheIdentifier": "custom-value-to-skip-babel-config-error",
                "caller": Object {
                  "__dangerous_rule_id": "expo-babel-loader",
                  "bundler": "webpack",
                  "mode": "production",
                  "platform": "ios",
                },
                "compact": true,
                "configFile": true,
                "inputSourceMap": true,
                "root": "packages/webpack-config/e2e/basic",
                "sourceMaps": true,
                "sourceType": "unambiguous",
              },
            },
          },
          Object {
            "test": /\\\\\\.worker\\\\\\.\\(js\\|mjs\\|ts\\)\\$/,
            "use": Object {
              "loader": "node_modules/worker-loader/dist/cjs.js",
            },
          },
          Object {
            "include": Array [
              "packages/webpack-config/e2e/basic",
              "packages/webpack-config/e2e/basic/node_modules/react-native-vector-icons",
              "packages/webpack-config/e2e/basic/node_modules/@expo/vector-icons",
            ],
            "test": /\\\\\\.\\(woff2\\?\\|eot\\|ttf\\|otf\\)\\$/,
            "use": Array [
              Object {
                "loader": "node_modules/url-loader/dist/cjs.js",
                "options": Object {
                  "esModule": false,
                  "limit": 50000,
                  "name": "./fonts/[name].[ext]",
                },
              },
            ],
          },
          Object {
            "test": /\\\\\\.\\(css\\)\\$/,
            "use": Array [
              "node_modules/style-loader/dist/cjs.js",
              "node_modules/css-loader/dist/cjs.js",
            ],
          },
          Object {
            "exclude": Array [
              /\\\\\\.\\(mjs\\|\\[jt\\]sx\\?\\)\\$/,
              /\\\\\\.html\\$/,
              /\\\\\\.json\\$/,
            ],
            "loader": "node_modules/file-loader/dist/cjs.js",
            "options": Object {
              "esModule": false,
              "name": "static/media/[name].[hash:8].[ext]",
            },
          },
        ],
      },
    ],
    "strictExportPresence": false,
  },
  "optimization": Object {
    "minimize": true,
    "minimizer": Array [
      Object {
        "options": Object {
          "cache": true,
          "cacheKeys": [Function],
          "exclude": undefined,
          "extractComments": true,
          "include": undefined,
          "minify": undefined,
          "parallel": true,
          "sourceMap": true,
          "terserOptions": Object {
            "compress": Object {
              "comparisons": false,
              "inline": 2,
              "warnings": false,
            },
            "mangle": Object {
              "safari10": true,
            },
            "output": Object {
              "ascii_only": true,
              "comments": false,
              "ecma": 5,
            },
            "parse": Object {
              "ecma": 8,
            },
          },
          "test": /\\\\\\.m\\?js\\(\\\\\\?\\.\\*\\)\\?\\$/i,
          "warningsFilter": [Function],
        },
      },
      Object {
        "deleteAssetsMap": Object {},
        "options": Object {
          "assetNameRegExp": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
          "assetProcessors": Array [
            Object {
              "phase": "compilation.optimize-chunk-assets",
              "processor": [Function],
              "regExp": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
            },
          ],
          "canPrint": undefined,
          "cssProcessor": [Function],
          "cssProcessorOptions": Object {
            "map": Object {
              "annotation": true,
              "inline": false,
            },
            "parser": [Function],
          },
          "cssProcessorPluginOptions": Object {},
        },
        "phaseAssetProcessors": Object {
          "compilation.optimize-assets": Array [],
          "compilation.optimize-chunk-assets": Array [
            Object {
              "phase": "compilation.optimize-chunk-assets",
              "processor": [Function],
              "regExp": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
            },
          ],
          "emit": Array [],
        },
        "pluginDescriptor": Object {
          "name": "OptimizeCssAssetsWebpackPlugin",
        },
      },
    ],
    "noEmitOnErrors": true,
    "nodeEnv": false,
    "runtimeChunk": true,
    "splitChunks": Object {
      "chunks": "all",
      "name": false,
    },
  },
  "output": Object {
    "chunkFilename": "static/js/[name].[contenthash:8].chunk.js",
    "devtoolModuleFilenameTemplate": [Function],
    "filename": "static/js/[name].[contenthash:8].js",
    "globalObject": "this",
    "path": "packages/webpack-config/e2e/basic/web-build",
    "publicPath": "/",
  },
  "plugins": Array [
    "CleanWebpackPlugin",
    "CopyPlugin",
    "HtmlWebpackPlugin",
    "InterpolateHtmlPlugin",
    "ExpoAppManifestWebpackPlugin",
    "ModuleNotFoundPlugin",
    "DefinePlugin",
    "LimitChunkCountPlugin",
  ],
  "resolve": Object {
    "extensions": Array [
      ".ios.ts",
      ".ios.tsx",
      ".ios.js",
      ".ios.jsx",
      ".native.ts",
      ".native.tsx",
      ".native.js",
      ".native.jsx",
      ".ts",
      ".tsx",
      ".js",
      ".jsx",
      ".json",
    ],
    "plugins": Array [
      Object {
        "apply": [Function],
        "bind": [Function],
        "forkTsCheckerOptions": [Function],
        "makePlugin": [Function],
        "moduleLoader": [Function],
        "topLevelLoader": Object {
          "apply": [Function],
        },
        "tsLoaderOptions": [Function],
      },
    ],
    "symlinks": false,
  },
  "resolveLoader": Object {
    "plugins": Array [
      Object {
        "apply": [Function],
      },
    ],
  },
  "target": "webworker",
}
`;

exports[`web web development 1`] = `
Object {
  "bail": false,
  "context": "packages/webpack-config/src",
  "devServer": Object {
    "after": [Function],
    "before": [Function],
    "clientLogLevel": "none",
    "compress": true,
    "contentBase": "packages/webpack-config/web-default",
    "contentBasePublicPath": "/",
    "disableHostCheck": true,
    "historyApiFallback": Object {
      "disableDotRule": true,
      "index": "/",
    },
    "host": "0.0.0.0",
    "hot": true,
    "https": false,
    "injectClient": false,
    "inline": true,
    "mimeTypes": undefined,
    "overlay": false,
    "proxy": undefined,
    "public": undefined,
    "publicPath": "/",
    "quiet": true,
    "sockHost": undefined,
    "sockPath": undefined,
    "sockPort": undefined,
    "transportMode": "ws",
    "watchContentBase": true,
  },
  "devtool": "cheap-module-source-map",
  "entry": Object {
    "app": Array [
      "node_modules/react-dev-utils/webpackHotDevClient.js",
      "packages/webpack-config/e2e/basic/index.js",
    ],
  },
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "parser": Object {
          "requireEnsure": false,
        },
      },
      Object {
        "oneOf": Array [
          Object {
            "exclude": "packages/webpack-config/web-default",
            "test": /\\\\\\.html\\$/,
            "use": Array [
              "node_modules/html-loader/dist/cjs.js",
            ],
          },
          Object {
            "test": /\\\\\\.\\(gif\\|jpe\\?g\\|png\\|svg\\)\\$/,
            "use": Object {
              "loader": "node_modules/url-loader/dist/cjs.js",
              "options": Object {
                "esModule": false,
                "limit": 1000,
                "name": "static/media/[name].[hash:8].[ext]",
              },
            },
          },
          Object {
            "include": [Function],
            "test": /\\\\\\.\\(mjs\\|\\[jt\\]sx\\?\\)\\$/,
            "use": Object {
              "loader": "node_modules/babel-loader/lib/index.js",
              "options": Object {
                "babelrc": false,
                "cacheCompression": false,
                "cacheDirectory": "packages/webpack-config/e2e/basic/.expo/web/cache/development/babel-loader",
                "cacheIdentifier": "custom-value-to-skip-babel-config-error",
                "caller": Object {
                  "__dangerous_rule_id": "expo-babel-loader",
                  "bundler": "webpack",
                  "mode": "development",
                  "platform": "web",
                },
                "compact": false,
                "configFile": true,
                "inputSourceMap": true,
                "root": "packages/webpack-config/e2e/basic",
                "sourceMaps": true,
                "sourceType": "unambiguous",
              },
            },
          },
          Object {
            "test": /\\\\\\.worker\\\\\\.\\(js\\|mjs\\|ts\\)\\$/,
            "use": Object {
              "loader": "node_modules/worker-loader/dist/cjs.js",
            },
          },
          Object {
            "include": Array [
              "packages/webpack-config/e2e/basic",
              "packages/webpack-config/e2e/basic/node_modules/react-native-vector-icons",
              "packages/webpack-config/e2e/basic/node_modules/@expo/vector-icons",
            ],
            "test": /\\\\\\.\\(woff2\\?\\|eot\\|ttf\\|otf\\)\\$/,
            "use": Array [
              Object {
                "loader": "node_modules/url-loader/dist/cjs.js",
                "options": Object {
                  "esModule": false,
                  "limit": 50000,
                  "name": "./fonts/[name].[ext]",
                },
              },
            ],
          },
          Object {
            "test": /\\\\\\.\\(css\\)\\$/,
            "use": Array [
              "node_modules/style-loader/dist/cjs.js",
              "node_modules/css-loader/dist/cjs.js",
            ],
          },
          Object {
            "exclude": Array [
              /\\\\\\.\\(mjs\\|\\[jt\\]sx\\?\\)\\$/,
              /\\\\\\.html\\$/,
              /\\\\\\.json\\$/,
            ],
            "loader": "node_modules/file-loader/dist/cjs.js",
            "options": Object {
              "esModule": false,
              "name": "static/media/[name].[hash:8].[ext]",
            },
          },
        ],
      },
    ],
    "strictExportPresence": false,
  },
  "node": Object {
    "child_process": "empty",
    "dgram": "empty",
    "dns": "mock",
    "fs": "empty",
    "http2": "empty",
    "module": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "static/js/[name].chunk.js",
    "devtoolModuleFilenameTemplate": [Function],
    "filename": "static/js/bundle.js",
    "globalObject": "this",
    "path": "packages/webpack-config/e2e/basic/web-build",
    "pathinfo": true,
    "publicPath": "/",
  },
  "plugins": Array [
    "HtmlWebpackPlugin",
    "InterpolateHtmlPlugin",
    "FaviconWebpackPlugin",
    "ModuleNotFoundPlugin",
    "DefinePlugin",
    "HotModuleReplacementPlugin",
    "WatchMissingNodeModulesPlugin",
    "ManifestPlugin",
  ],
  "resolve": Object {
    "alias": Object {
      "react-native$": "react-native-web",
      "react-native/Libraries/Components/View/ViewStylePropTypes$": "react-native-web/dist/exports/View/ViewStylePropTypes",
      "react-native/Libraries/EventEmitter/NativeEventEmitter$": "react-native-web/dist/vendor/react-native/NativeEventEmitter",
      "react-native/Libraries/EventEmitter/RCTDeviceEventEmitter$": "react-native-web/dist/vendor/react-native/NativeEventEmitter/RCTDeviceEventEmitter",
      "react-native/Libraries/Image/AssetSourceResolver$": "expo-asset/build/AssetSourceResolver",
      "react-native/Libraries/Image/assetPathUtils$": "expo-asset/build/Image/assetPathUtils",
      "react-native/Libraries/Image/resolveAssetSource$": "expo-asset/build/resolveAssetSource",
      "react-native/Libraries/vendor/emitter/EventEmitter$": "react-native-web/dist/vendor/react-native/emitter/EventEmitter",
      "react-native/Libraries/vendor/emitter/EventSubscriptionVendor$": "react-native-web/dist/vendor/react-native/emitter/EventSubscriptionVendor",
    },
    "extensions": Array [
      ".web.ts",
      ".web.tsx",
      ".web.mjs",
      ".web.js",
      ".web.jsx",
      ".ts",
      ".tsx",
      ".mjs",
      ".js",
      ".jsx",
      ".json",
      ".wasm",
    ],
    "plugins": Array [
      Object {
        "apply": [Function],
        "bind": [Function],
        "forkTsCheckerOptions": [Function],
        "makePlugin": [Function],
        "moduleLoader": [Function],
        "topLevelLoader": Object {
          "apply": [Function],
        },
        "tsLoaderOptions": [Function],
      },
    ],
    "symlinks": false,
  },
  "resolveLoader": Object {
    "plugins": Array [
      Object {
        "apply": [Function],
      },
    ],
  },
}
`;

exports[`web web production 1`] = `
Object {
  "bail": true,
  "context": "packages/webpack-config/src",
  "devtool": "source-map",
  "entry": Object {
    "app": Array [
      "packages/webpack-config/e2e/basic/index.js",
    ],
  },
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "parser": Object {
          "requireEnsure": false,
        },
      },
      Object {
        "oneOf": Array [
          Object {
            "exclude": "packages/webpack-config/web-default",
            "test": /\\\\\\.html\\$/,
            "use": Array [
              "node_modules/html-loader/dist/cjs.js",
            ],
          },
          Object {
            "test": /\\\\\\.\\(gif\\|jpe\\?g\\|png\\|svg\\)\\$/,
            "use": Object {
              "loader": "node_modules/url-loader/dist/cjs.js",
              "options": Object {
                "esModule": false,
                "limit": 1000,
                "name": "static/media/[name].[hash:8].[ext]",
              },
            },
          },
          Object {
            "include": [Function],
            "test": /\\\\\\.\\(mjs\\|\\[jt\\]sx\\?\\)\\$/,
            "use": Object {
              "loader": "node_modules/babel-loader/lib/index.js",
              "options": Object {
                "babelrc": false,
                "cacheCompression": false,
                "cacheDirectory": "packages/webpack-config/e2e/basic/.expo/web/cache/production/babel-loader",
                "cacheIdentifier": "custom-value-to-skip-babel-config-error",
                "caller": Object {
                  "__dangerous_rule_id": "expo-babel-loader",
                  "bundler": "webpack",
                  "mode": "production",
                  "platform": "web",
                },
                "compact": true,
                "configFile": true,
                "inputSourceMap": true,
                "root": "packages/webpack-config/e2e/basic",
                "sourceMaps": true,
                "sourceType": "unambiguous",
              },
            },
          },
          Object {
            "test": /\\\\\\.worker\\\\\\.\\(js\\|mjs\\|ts\\)\\$/,
            "use": Object {
              "loader": "node_modules/worker-loader/dist/cjs.js",
            },
          },
          Object {
            "include": Array [
              "packages/webpack-config/e2e/basic",
              "packages/webpack-config/e2e/basic/node_modules/react-native-vector-icons",
              "packages/webpack-config/e2e/basic/node_modules/@expo/vector-icons",
            ],
            "test": /\\\\\\.\\(woff2\\?\\|eot\\|ttf\\|otf\\)\\$/,
            "use": Array [
              Object {
                "loader": "node_modules/url-loader/dist/cjs.js",
                "options": Object {
                  "esModule": false,
                  "limit": 50000,
                  "name": "./fonts/[name].[ext]",
                },
              },
            ],
          },
          Object {
            "test": /\\\\\\.\\(css\\)\\$/,
            "use": Array [
              "node_modules/style-loader/dist/cjs.js",
              "node_modules/css-loader/dist/cjs.js",
            ],
          },
          Object {
            "exclude": Array [
              /\\\\\\.\\(mjs\\|\\[jt\\]sx\\?\\)\\$/,
              /\\\\\\.html\\$/,
              /\\\\\\.json\\$/,
            ],
            "loader": "node_modules/file-loader/dist/cjs.js",
            "options": Object {
              "esModule": false,
              "name": "static/media/[name].[hash:8].[ext]",
            },
          },
        ],
      },
    ],
    "strictExportPresence": false,
  },
  "node": Object {
    "child_process": "empty",
    "dgram": "empty",
    "dns": "mock",
    "fs": "empty",
    "http2": "empty",
    "module": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "optimization": Object {
    "minimize": true,
    "minimizer": Array [
      Object {
        "options": Object {
          "cache": true,
          "cacheKeys": [Function],
          "exclude": undefined,
          "extractComments": true,
          "include": undefined,
          "minify": undefined,
          "parallel": true,
          "sourceMap": true,
          "terserOptions": Object {
            "compress": Object {
              "comparisons": false,
              "inline": 2,
              "warnings": false,
            },
            "mangle": Object {
              "safari10": true,
            },
            "output": Object {
              "ascii_only": true,
              "comments": false,
              "ecma": 5,
            },
            "parse": Object {
              "ecma": 8,
            },
          },
          "test": /\\\\\\.m\\?js\\(\\\\\\?\\.\\*\\)\\?\\$/i,
          "warningsFilter": [Function],
        },
      },
      Object {
        "deleteAssetsMap": Object {},
        "options": Object {
          "assetNameRegExp": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
          "assetProcessors": Array [
            Object {
              "phase": "compilation.optimize-chunk-assets",
              "processor": [Function],
              "regExp": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
            },
          ],
          "canPrint": undefined,
          "cssProcessor": [Function],
          "cssProcessorOptions": Object {
            "map": Object {
              "annotation": true,
              "inline": false,
            },
            "parser": [Function],
          },
          "cssProcessorPluginOptions": Object {},
        },
        "phaseAssetProcessors": Object {
          "compilation.optimize-assets": Array [],
          "compilation.optimize-chunk-assets": Array [
            Object {
              "phase": "compilation.optimize-chunk-assets",
              "processor": [Function],
              "regExp": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
            },
          ],
          "emit": Array [],
        },
        "pluginDescriptor": Object {
          "name": "OptimizeCssAssetsWebpackPlugin",
        },
      },
    ],
    "noEmitOnErrors": true,
    "nodeEnv": false,
    "runtimeChunk": true,
    "splitChunks": Object {
      "chunks": "all",
      "name": false,
    },
  },
  "output": Object {
    "chunkFilename": "static/js/[name].[contenthash:8].chunk.js",
    "devtoolModuleFilenameTemplate": [Function],
    "filename": "static/js/[name].[contenthash:8].js",
    "globalObject": "this",
    "path": "packages/webpack-config/e2e/basic/web-build",
    "publicPath": "/",
  },
  "plugins": Array [
    "CleanWebpackPlugin",
    "CopyPlugin",
    "HtmlWebpackPlugin",
    "InterpolateHtmlPlugin",
    "FaviconWebpackPlugin",
    "ApplePwaWebpackPlugin",
    "ChromeIconsWebpackPlugin",
    "ModuleNotFoundPlugin",
    "DefinePlugin",
    "MiniCssExtractPlugin",
    "ManifestPlugin",
  ],
  "resolve": Object {
    "alias": Object {
      "react-native$": "react-native-web",
      "react-native/Libraries/Components/View/ViewStylePropTypes$": "react-native-web/dist/exports/View/ViewStylePropTypes",
      "react-native/Libraries/EventEmitter/NativeEventEmitter$": "react-native-web/dist/vendor/react-native/NativeEventEmitter",
      "react-native/Libraries/EventEmitter/RCTDeviceEventEmitter$": "react-native-web/dist/vendor/react-native/NativeEventEmitter/RCTDeviceEventEmitter",
      "react-native/Libraries/Image/AssetSourceResolver$": "expo-asset/build/AssetSourceResolver",
      "react-native/Libraries/Image/assetPathUtils$": "expo-asset/build/Image/assetPathUtils",
      "react-native/Libraries/Image/resolveAssetSource$": "expo-asset/build/resolveAssetSource",
      "react-native/Libraries/vendor/emitter/EventEmitter$": "react-native-web/dist/vendor/react-native/emitter/EventEmitter",
      "react-native/Libraries/vendor/emitter/EventSubscriptionVendor$": "react-native-web/dist/vendor/react-native/emitter/EventSubscriptionVendor",
    },
    "extensions": Array [
      ".web.ts",
      ".web.tsx",
      ".web.mjs",
      ".web.js",
      ".web.jsx",
      ".ts",
      ".tsx",
      ".mjs",
      ".js",
      ".jsx",
      ".json",
      ".wasm",
    ],
    "plugins": Array [
      Object {
        "apply": [Function],
        "bind": [Function],
        "forkTsCheckerOptions": [Function],
        "makePlugin": [Function],
        "moduleLoader": [Function],
        "topLevelLoader": Object {
          "apply": [Function],
        },
        "tsLoaderOptions": [Function],
      },
    ],
    "symlinks": false,
  },
  "resolveLoader": Object {
    "plugins": Array [
      Object {
        "apply": [Function],
      },
    ],
  },
}
`;
